File: package/go/src/scope_manager.go
package src

// ScopeManager manages variable scopes in a lexical scoping system
type ScopeManager struct {
	// Stack of scopes, where each scope is a map of variable names to values
	// The first scope is always the global scope
	scopes []map[string]int
}

// NewScopeManager creates a new scope manager with a global scope
func NewScopeManager() *ScopeManager {
	return &ScopeManager{
		scopes: []map[string]int{make(map[string]int)},
	}
}

// EnterScope creates a new scope and makes the current scope one level deeper
func (sm *ScopeManager) EnterScope() {
	// Add a new dictionary to the stack to represent the inner scope
	sm.scopes = append(sm.scopes, make(map[string]int))
}

// ExitScope destroys the current scope and returns to the one level up scope.
// If attempting to delete the top-level global scope, it does nothing.
func (sm *ScopeManager) ExitScope() {
	// Do nothing if the global scope is the only one
	if len(sm.scopes) > 1 {
		sm.scopes = sm.scopes[:len(sm.scopes)-1]
	}
}

// SetVariable saves the value with the variable name as the key in the current scope
func (sm *ScopeManager) SetVariable(name string, value int) {
	// Set the variable in the current (innermost) scope
	sm.scopes[len(sm.scopes)-1][name] = value
}

// GetVariable searches for the variable name from the current scope in order,
// and returns its value if found.
// Returns nil if it doesn't exist in any scope.
func (sm *ScopeManager) GetVariable(name string) *int {
	// Search from the innermost scope in order and return the first variable found (lexical scope behavior)
	for i := len(sm.scopes) - 1; i >= 0; i-- {
		if value, exists := sm.scopes[i][name]; exists {
			return &value
		}
	}
	return nil
}

File: package/php/src/ScopeManager.php
<?php

declare(strict_types=1);

namespace ProgramKnock;

class ScopeManager
{
    /**
     * スコープのスタック（各スコープは変数名=>値の配列）
     * @var array<array<string, int>>
     */
    private array $scopeStack;

    public function __construct()
    {
        // グローバルスコープから開始
        $this->scopeStack = [[]];
    }

    /**
     * 新しいスコープに入る
     */
    public function enterScope(): void
    {
        $this->scopeStack[] = [];
    }

    /**
     * 現在のスコープから出る
     */
    public function exitScope(): void
    {
        // グローバルスコープの場合は何もしない
        if (count($this->scopeStack) > 1) {
            array_pop($this->scopeStack);
        }
    }

    /**
     * 現在のスコープに変数を設定
     */
    public function setVariable(string $name, int $value): void
    {
        $currentScopeIndex = count($this->scopeStack) - 1;
        $this->scopeStack[$currentScopeIndex][$name] = $value;
    }

    /**
     * 変数の値を取得（内側から外側のスコープに向かって検索）
     */
    public function getVariable(string $name): ?int
    {
        // 内側のスコープから外側のスコープに向かって検索
        for ($i = count($this->scopeStack) - 1; $i >= 0; $i--) {
            if (isset($this->scopeStack[$i][$name])) {
                return $this->scopeStack[$i][$name];
            }
        }

        return null;
    }

    /**
     * 現在のスコープレベルを取得（テスト用）
     */
    public function getCurrentScopeLevel(): int
    {
        return count($this->scopeStack) - 1;
    }

    /**
     * 現在のスコープの変数一覧を取得（テスト用）
     *
     * @return array<string, int>
     */
    public function getCurrentScopeVariables(): array
    {
        return $this->scopeStack[count($this->scopeStack) - 1];
    }
}
